transcribers/mqtt.py

```diff
@@ -131,28 +133,53 @@ def _mqtt_to_ipal(self, request, pkt):
             data=data,
         )
 
+        # depending on the message type, add the message to the request queue or match it to a request in the queue
+        ipal_message._add_to_request_queue = (type in MQTTProtocol.REQUEST_COMMANDS)
+        ipal_message._match_to_requests = (type in MQTTProtocol.RESPONSE_COMMANDS)
+
+        return ipal_message
+
     def match_response(self, requests, response):
         remove_from_queue = []
 
-        # preemptively reduce the amount of checks we need to do
-        match_types = MQTTProtocol.command_response(response)
-        if len(match_types) == 0:
-            return []
-
-        # then only check relevant types
         for request in requests:
-            if request.type in match_types:
-                if (request.src, request.dest) == (response.dest, response.src):
-                    if request.type in {MQTTProtocol.CONNECT, MQTTProtocol.PINGREQ}:
-                        response.responds_to.append(request.id)
-                        remove_from_queue.append(request)
-
-                    elif (
-                        request.type
-                        in MQTTProtocol.TOPIC_COMMANDS | MQTTProtocol.PUBLISH_COMMANDS
-                        and request.keys() == response.keys()
-                    ):
-                        response.responds_to.append(request.id)
-                        remove_from_queue.append(request)
+            if response.type == MQTTProtocol.PUBACK:
+                if request.type == MQTTProtocol.PUBLISH and request.src == response.dest:
+                    response.responds_to.append(request.id)
+                    remove_from_queue.append(request)
+                    break
```

Comment:
When we send 2 PUBACK (possibly malicious), the second PUBACK will not respond_to the PUBLISH. Is this intended behaviour?

Reply:
Yes this is intended. Another approach to circumvent this problem would be to keep a list of the last N packets which could be responded to. Whenever a new packet is seen, one would always remove the oldest packet from this list instead of removing a packet whenever it is responded to. However, this would require finding an N which is large enough such that no important packets are removed from the list and which at the same time does not cause too much overhead. I would say this is a design choice.